File: lib/domain/shared/gateways/user_lookup_gateway.rb
# frozen_string_literal: true

module Domain
  module Shared
    module Gateways
      # Interactor が User モデルを直接参照しないためのゲートウェイ。
      # 実装は Adapters::Shared::Gateways::UserActiveRecordGateway（CompositionRoot で生成）
      module UserLookupGateway
        # @return [Domain::Shared::Dtos::User] Policy / DeletionUndo の actor 用
        def find(user_id)
          raise NotImplementedError, "#{self.class}#find"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/shared/gateways/user_lookup_gateway.rs
use crate::shared::user::User;

/// Ruby: `Domain::Shared::Gateways::UserLookupGateway`
pub trait UserLookupGateway: Send + Sync {
    /// Ruby: `#find(user_id)` — `Domain::Shared::Dtos::User`
    fn find(&self, user_id: i64) -> User;
}
